### DISCUSSION TITLE
Perforce Error in Project /& Git Repo

### DISCUSSION THREAD

**Discussion by dkr91** [oracle/opengrok#4572] [Q&A]
**Describe the bug**
```log
WARNING [OpenGrok-index-db-000] org.opengrok.indexer.util.Executor.exec Non-zero exit status 1 from command [/usr/bin/p4 dirs $'*'] in directory '/opengrok/src/release-2x': Perforce client error:
        Connect to server failed; check $P4PORT.
        TCP connect to perforce:1666 failed.
        Name or service not known
```

If you provide the exact version of OpenGrok, JDK used, OS(and its version) used and Tomcat(or your webapp server) used it can help figuring out an environment issue.
For performance problem OS and JDK tunables might be needed. For SCM problems also version of SCM is helpfull (e.g. some mercurial versions have issues, some not).

Opengrok Docker 1.13.4/7
Indexer Options: 
```
depth 10 --reduceSegmentCount --leadingWildCards on -G -S --renamedHistory on -p release-2x -i d:LanguageFiles -i d:versioncache -i d:work-cache -i *.a -i *.bz2 -i *.class -i *.dll -i *.exe -i *.gz -i *.jar -i *.lib -i *.obj -i *.pdb -i *.so -i *.tar -i *.war -i *.xz -i *.zip
```

**To Reproduce**
Steps to reproduce the behavior: run Opengrok container.

**Expected behavior**
Perforce TCP not even trying because there are no Perforce Projects configured

**Comment by vladak** [comment]
`Name or service not known` probably means that the lookup of the Perforce server hostname failed. The container needs to have the naming set up so that it can perform the lookup.

**Reply by dkr91** [reply]
Why is it using Perforce if I only have Git and SVN?

**Reply by vladak** [reply]
This is likely part of the repository scan/invalidation. Perforce is installed in the Docker image by default. Maybe there is smarter way to perform the Perforce repository detection to avoid the noise.

**Reply by cyberplant** [reply]
Are these warnings harmless?

I'm being hit by this and I'm evaluating options.. 

* Disabling Perforce by env_var or config if it's possible
* Run the p4d daemon inside the container

I honestly never heard of Perforce before, it's the first time for me, so I rather have it disabled 😄

**Reply by subramaniam** [reply]
I see the same issue:

` docker version
Client:
 Version:           27.1.1
 API version:       1.46
 Go version:        go1.21.12
 Git commit:        6312585
 Built:             Tue Jul 23 19:57:57 2024
 OS/Arch:           windows/amd64
 Context:           desktop-linux

Server: Docker Desktop 4.33.0 (160616)
 Engine:
  Version:          27.1.1
  API version:      1.46 (minimum version 1.24)
  Go version:       go1.21.12
  Git commit:       cc13f95
  Built:            Tue Jul 23 19:57:19 2024
  OS/Arch:          linux/amd64
  Experimental:     false
 containerd:
  Version:          1.7.19
  GitCommit:        2bf793ef6dc9a18e00cb12efb64355c2c9d5eb41
 runc:
  Version:          1.7.19
  GitCommit:        v1.1.13-0-g58aa920
 docker-init:
  Version:          0.19.0
  GitCommit:        de40ad0`

The sources that I have setup do not have perforce and yet I see logs 
`
2024-07-28 20:29:23 29-Jul-2024 01:29:23.094 WARNING [OpenGrok-index-db-115] org.opengrok.indexer.util.Executor.exec Non-zero exit status 1 from command [/usr/bin/p4 files $'*'] in directory '/opengrok/src/qnx/utils/ipc_trace_logger': Perforce client error:
2024-07-28 20:29:23 Connect to server failed; check $P4PORT.
`

[full-logs.zip](https://github.com/user-attachments/files/16406215/full-logs.zip)

Please let me know if any further logs are required.

**Comment by hkw2pg1** [comment]
I worked around this by "docker exec" inside the container and uninstall all Perforce related packages
ii  helix-cli                   2024.2-2726408~jammy              amd64        Command-line client for the Helix Versioning Engine
ii  helix-cli-base              2024.2-2726408~jammy              amd64        Command-line client for the Helix Versioning Engine
ii  helix-p4d                   2024.2-2726408~jammy              amd64        Helix Core Server (p4d)
ii  helix-p4d-base              2024.2-2726408~jammy              amd64        Helix Core Server (p4d)
ii  helix-p4d-base-24.2         2024.2-2726408~jammy              amd64        Helix Core Server (p4d)
ii  helix-p4dctl                2024.2-2726408~jammy              amd64        Helper utility for managing Helix services

After that the repeated perforce warning does not appear anymore, instead it will just show once the P4 is not available. 
10-Apr-2025 08:00:12.588 WARNING [OpenGrok-index-db-197] org.opengrok.indexer.history.PerforceRepository.isP4Working Command '/usr/bin/p4' does not work. Some operations with Perforce repositories will fail as a result.

This will improve the efficiency of indexer as it no longer keep calling /usr/bin/p4 when indexing Git repos

**Comment by SebastienTolron** [comment]
Ok so : 

https://github.com/oracle/opengrok/pull/4987

Made with Claude.

Feel free to take it and contribute. I just wanted to help. Dont want to subscribe to OCA etc...
